<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>polygraph package &#8212; polygraph unknown documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=0ca6144b" />
    <script src="../_static/documentation_options.js?v=3f1b9271"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="polygraph" href="modules.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="polygraph-package">
<h1>polygraph package<a class="headerlink" href="#polygraph-package" title="Link to this heading">¶</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">¶</a></h2>
</section>
<section id="module-polygraph.classifier">
<span id="polygraph-classifier-module"></span><h2>polygraph.classifier module<a class="headerlink" href="#module-polygraph.classifier" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.classifier.groupwise_svm">
<span class="sig-prename descclassname"><span class="pre">polygraph.classifier.</span></span><span class="sig-name descname"><span class="pre">groupwise_svm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ad</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_kernel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/classifier.html#groupwise_svm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.classifier.groupwise_svm" title="Link to this definition">¶</a></dt>
<dd><p>Train an SVM to distinguish between each non-reference group and the reference group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ad</strong> (<em>anndata.AnnData</em>) – Anndata object containing sequence embeddings
of shape (n_seqs x n_vars)</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID of group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column in .obs containing group ID</p></li>
<li><p><strong>cv</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of cross-validation folds</p></li>
<li><p><strong>is_kernel</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Whether ad.X is a symmetric kernel matrix</p></li>
<li><p><strong>max_iter</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Maximum number of iterations for SVM</p></li>
<li><p><strong>use_pca</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Whether to use PCA distances</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Modified anndata object containing each</dt><dd><p>sequence’s predicted label in .obs, as well as SVM
performance metrics in ad.uns[“svm_performance”]</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ad (anndata.AnnData)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="polygraph-embedding-module">
<h2>polygraph.embedding module<a class="headerlink" href="#polygraph-embedding-module" title="Link to this heading">¶</a></h2>
</section>
<section id="module-polygraph.evolve">
<span id="polygraph-evolve-module"></span><h2>polygraph.evolve module<a class="headerlink" href="#module-polygraph.evolve" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.evolve.evolve">
<span class="sig-prename descclassname"><span class="pre">polygraph.evolve.</span></span><span class="sig-name descname"><span class="pre">evolve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_last_layers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">512</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">task</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/evolve.html#evolve"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.evolve.evolve" title="Link to this definition">¶</a></dt>
<dd><p>Directed evolution with an additional goal to increase similarity to
reference sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start_seq</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Start sequence</p></li>
<li><p><strong>reference_seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – Reference sequences</p></li>
<li><p><strong>iter</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of iterations</p></li>
<li><p><strong>model</strong> (<em>nn.Sequential</em>) – Torch sequential model</p></li>
<li><p><strong>k</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – k-mer length for k-mer embedding.</p></li>
<li><p><strong>drop_last_layers</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of terminal layers to drop from the
model for model embedding.</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size for inference</p></li>
<li><p><strong>device</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Index of device to use for inference</p></li>
<li><p><strong>task</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Model output head. If None, average all heads.</p></li>
<li><p><strong>alpha</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Relative weight for similarity</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Optimized sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>best_seq (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)">str</a>)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.input">
<span id="polygraph-input-module"></span><h2>polygraph.input module<a class="headerlink" href="#module-polygraph.input" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.input.download_gtex_tpm">
<span class="sig-prename descclassname"><span class="pre">polygraph.input.</span></span><span class="sig-name descname"><span class="pre">download_gtex_tpm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">download_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/runner/work/polygraph/polygraph/src/polygraph/resources/gtex'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/input.html#download_gtex_tpm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.input.download_gtex_tpm" title="Link to this definition">¶</a></dt>
<dd><p>Download per-tissue TPM values from GTEX.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>download_dir</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to directory in which to download file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Path to downloaded local file</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.input.download_jaspar">
<span class="sig-prename descclassname"><span class="pre">polygraph.input.</span></span><span class="sig-name descname"><span class="pre">download_jaspar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">family</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'vertebrates'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">download_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/runner/work/polygraph/polygraph/src/polygraph/resources/jaspar'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/input.html#download_jaspar"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.input.download_jaspar" title="Link to this definition">¶</a></dt>
<dd><p>Download and read the JASPAR database of TF motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>family</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – JASPAR family. one of “fungi”, “insects”, “nematodes”,
“plants”, “urochordates”, “vertebrates”</p></li>
<li><p><strong>download_dir</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to directory in which to download motifs</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Path to downloaded local file</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.input.load_gtex_tpm">
<span class="sig-prename descclassname"><span class="pre">polygraph.input.</span></span><span class="sig-name descname"><span class="pre">load_gtex_tpm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">download_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/runner/work/polygraph/polygraph/src/polygraph/resources/gtex'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/input.html#load_gtex_tpm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.input.load_gtex_tpm" title="Link to this definition">¶</a></dt>
<dd><p>Load per-tissue TPM values from GTEX.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>download_dir</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to directory in which to download file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>TPM matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.input.read_meme_file">
<span class="sig-prename descclassname"><span class="pre">polygraph.input.</span></span><span class="sig-name descname"><span class="pre">read_meme_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/input.html#read_meme_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.input.read_meme_file" title="Link to this definition">¶</a></dt>
<dd><p>Read a motif database in MEME format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – path to MEME file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of pymemesuite.common.Motif objects
bg (pymemesuite.common.Background): Background distribution</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>motifs (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.input.read_seqs">
<span class="sig-prename descclassname"><span class="pre">polygraph.input.</span></span><span class="sig-name descname"><span class="pre">read_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/input.html#read_seqs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.input.read_seqs" title="Link to this definition">¶</a></dt>
<dd><p>Read sequences and group labels into a dataframe. This creates the input
dataframe for all subsequent analyses.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – path to a text file containing no header. If incl_ids=True,</p></li>
<li><p><strong>contain</strong> (<em>the first column should contain IDs and the next two columns should</em>)</p></li>
<li><p><strong>incl_ids=False</strong> (<em>sequence and group label. If</em>)</p></li>
<li><p><strong>should</strong> (<em>the first two columns</em>)</p></li>
<li><p><strong>label.</strong> (<em>contain sequence and group</em>)</p></li>
<li><p><strong>sep</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column separator</p></li>
<li><p><strong>incl_ids</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Whether the first column corresponds to sequence IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Pandas dataframe with columns Sequence, Group</dt><dd><p>and a unique index.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>df (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.likelihood">
<span id="polygraph-likelihood-module"></span><h2>polygraph.likelihood module<a class="headerlink" href="#module-polygraph.likelihood" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="polygraph.likelihood.CharDataset">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">polygraph.likelihood.</span></span><span class="sig-name descname"><span class="pre">CharDataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/likelihood.html#CharDataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.likelihood.CharDataset" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="polygraph.likelihood.CharDataset.encode">
<span class="sig-name descname"><span class="pre">encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/likelihood.html#CharDataset.encode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.likelihood.CharDataset.encode" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.likelihood.compute_likelihood">
<span class="sig-prename descclassname"><span class="pre">polygraph.likelihood.</span></span><span class="sig-name descname"><span class="pre">compute_likelihood</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/likelihood.html#compute_likelihood"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.likelihood.compute_likelihood" title="Link to this definition">¶</a></dt>
<dd><p>Function to compute the log-likelihood of each sequence in the
given list using the hyenaDNA model pretrained on the human genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – DNA sequence, list of DNA sequences
or a dataframe containing sequences in the column “Sequence”.</p></li>
<li><p><strong>model</strong> (<em>ConvLMHead</em>) – HyenaDNA model</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size for inference</p></li>
<li><p><strong>num_workers</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of workers for inference dataloader</p></li>
<li><p><strong>device</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Device ID for inference</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Log-likelihoods for each sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>LL (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.likelihood.load_hyenadna">
<span class="sig-prename descclassname"><span class="pre">polygraph.likelihood.</span></span><span class="sig-name descname"><span class="pre">load_hyenadna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hyena_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ckpt_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hyenadna-small-32k-seqlen'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/likelihood.html#load_hyenadna"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.likelihood.load_hyenadna" title="Link to this definition">¶</a></dt>
<dd><p>Loads the pretrained hyenaDNA foundation model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hyena_path</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to the cloned hyenaDNA repo. The repo must be cloned with
the recurse-submodules flag. See installation instructions at
<a class="reference external" href="https://github.com/HazyResearch/hyena-dna/tree/main">https://github.com/HazyResearch/hyena-dna/tree/main</a>.</p></li>
<li><p><strong>ckpt_dir</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to directory in which to download the model</p></li>
<li><p><strong>model</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of the foundation model to download. See
<a class="reference external" href="https://github.com/HazyResearch/hyena-dna/tree/main">https://github.com/HazyResearch/hyena-dna/tree/main</a> for options.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pretrained HyenaDNA model</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>model (ConvLMHeadModel)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.models">
<span id="polygraph-models-module"></span><h2>polygraph.models module<a class="headerlink" href="#module-polygraph.models" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.batch">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">batch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#batch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.batch" title="Link to this definition">¶</a></dt>
<dd><p>Pad sequences to a constant length and split them into batches to pass to a model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequences</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of DNA sequences</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>sequence batch generator</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.cell_type_specificity">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">cell_type_specificity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_target_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">off_target_cols</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#cell_type_specificity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.cell_type_specificity" title="Link to this definition">¶</a></dt>
<dd><p>Calculate cell type specificity from predicted or measured output</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Dataframe containing sequence predictions</p></li>
<li><p><strong>on_target</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing predictions in on-target cell type</p></li>
<li><p><strong>off_target</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – Columns containing predictions in off-target cell types.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>seqs with additional columns mingap, maxgap and meangap,</dt><dd><p>reporting 3 measures of cell type specificity for each sequence.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.enformer_embed">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">enformer_embed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#enformer_embed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.enformer_embed" title="Link to this definition">¶</a></dt>
<dd><p>Embed a batch of sequences using pretrained or fine-tuned enformer</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequences</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of sequences</p></li>
<li><p><strong>model</strong> (<em>Enformer</em>) – pre-trained or fine-tuned enformer model</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>np.array of shape (n_seqs x 3072)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.get_embeddings">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">get_embeddings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_last_layers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">swapaxes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#get_embeddings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.get_embeddings" title="Link to this definition">¶</a></dt>
<dd><p>Get model embeddings for all sequences in a dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – List of sequences or dataframe
containing sequences in the column “Sequence”.</p></li>
<li><p><strong>model</strong> (<em>nn.Sequential</em>) – trained model</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size for inference</p></li>
<li><p><strong>drop_last_layers</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of terminal layers to drop to get embeddings</p></li>
<li><p><strong>device</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – ID of GPU to perform inference.</p></li>
<li><p><strong>swapaxes</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If true, batches will be of shape (N, 4, L).
Otherwise, shape will be (N, L, 4).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>np.array of shape (n_seqs x n_features)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.ism_score">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">ism_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">task</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#ism_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.ism_score" title="Link to this definition">¶</a></dt>
<dd><p>Get base-level importance scores for given sequence(s) using ISM</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – List of sequences or dataframe
containing sequences in the column “Sequence”.</p></li>
<li><p><strong>model</strong> (<em>nn.Sequential</em>) – trained model</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size for inference</p></li>
<li><p><strong>device</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – ID of GPU to perform inference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame of shape (n_seqs x n_outputs)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.load_enformer">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">load_enformer</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#load_enformer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.load_enformer" title="Link to this definition">¶</a></dt>
<dd><p>Load pre-trained enformer model</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Pretrained model</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>(Enformer)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.load_nucleotide_transformer">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">load_nucleotide_transformer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'InstaDeepAI/nucleotide-transformer-2.5b-multi-species'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#load_nucleotide_transformer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.load_nucleotide_transformer" title="Link to this definition">¶</a></dt>
<dd><p>Load pre-trained nucleotide transformer model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>model</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of pretrained model to download</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pre-trained model
tokenizer (): Class to convert sequences to tokens</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>model (EsmForMaskedLM)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.nucleotide_transformer_embed">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">nucleotide_transformer_embed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tokenizer</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#nucleotide_transformer_embed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.nucleotide_transformer_embed" title="Link to this definition">¶</a></dt>
<dd><p>Embed a batch of sequences using the pre-trained nucleotide transformer model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequences</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of sequences</p></li>
<li><p><strong>model</strong> – pre-trained nucleotide transformer model</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>np.array of shape (n_seqs x n_features)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.predict">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.predict" title="Link to this definition">¶</a></dt>
<dd><p>Predict sequence properties using a sequence-to-function model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – List of sequences or dataframe
containing sequences in the column “Sequence”.</p></li>
<li><p><strong>model</strong> (<em>nn.Sequential</em>) – trained model</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size for inference</p></li>
<li><p><strong>device</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – ID of GPU to perform inference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of shape (n_seqs x n_outputs)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.robustness">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">robustness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">task</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aggfunc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#robustness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.robustness" title="Link to this definition">¶</a></dt>
<dd><p>Get robustness scores for given sequence(s) using ISM</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – List of sequences or dataframe
containing sequences in the column “Sequence”.</p></li>
<li><p><strong>model</strong> (<em>nn.Sequential</em>) – trained model</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Batch size for inference</p></li>
<li><p><strong>device</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – ID of GPU to perform inference.</p></li>
<li><p><strong>aggfunc</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Either ‘mean’ or ‘max’. Determines how to aggregate the
effect of all possible single-base mutations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame of shape (n_seqs x n_outputs)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.models.sequential_embed">
<span class="sig-prename descclassname"><span class="pre">polygraph.models.</span></span><span class="sig-name descname"><span class="pre">sequential_embed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_last_layers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">swapaxes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/models.html#sequential_embed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.models.sequential_embed" title="Link to this definition">¶</a></dt>
<dd><p>Embed a batch of sequences using a torch.nn.Sequential model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of sequences</p></li>
<li><p><strong>model</strong> (<em>nn.Sequential</em>) – trained model</p></li>
<li><p><strong>drop_last_layers</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of terminal layers to drop to get embeddings</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>np.array of shape (n_seqs x n_features)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.motifs">
<span id="polygraph-motifs-module"></span><h2>polygraph.motifs module<a class="headerlink" href="#module-polygraph.motifs" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.get_motif_pairs">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">get_motif_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sites</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#get_motif_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.get_motif_pairs" title="Link to this definition">¶</a></dt>
<dd><p>List the pairs of motifs present in each sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>sites</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing FIMO output.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing all motif pairs in each
sequence with their orientation and distance.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pairs (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.motif_frequencies">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">motif_frequencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#motif_frequencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.motif_frequencies" title="Link to this definition">¶</a></dt>
<dd><p>Count frequency of occurrence of motifs in a list of sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sites</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – Output of <cite>scan</cite> function</p></li>
<li><p><strong>normalize</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Whether to normalize the resulting count matrix
to correct for sequence length</p></li>
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing DNA sequences.
Needed if normalize=True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Count matrix with rows = sequences and columns = motifs</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cts (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.motif_pair_differential_abundance">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">motif_pair_differential_abundance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_pairs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#motif_pair_differential_abundance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.motif_pair_differential_abundance" title="Link to this definition">¶</a></dt>
<dd><p>Compare the rate of occurence of pairwise combinations of motifs between groups</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif_pairs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing the ouptut of
get_motif_pairs.</p></li>
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing sequences</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID of group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column in <cite>seqs</cite> containing group IDs</p></li>
<li><p><strong>max_prop_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Limit to combinations with this proportion in
at least one group.</p></li>
<li><p><strong>min_prop_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – Limit to combinations with this proportion in
in all groups.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Pandas dataframe containing FDR-corrected significance</dt><dd><p>testing results for the occurrence of pairwise combinations between groups</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>res (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.motif_pair_differential_distance">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">motif_pair_differential_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_pairs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#motif_pair_differential_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.motif_pair_differential_distance" title="Link to this definition">¶</a></dt>
<dd><p>Compare the distance between all motif pairs across groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif_pairs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing the ouptut of
get_motif_pairs.</p></li>
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing sequences</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID of group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column in <cite>seqs</cite> containing group IDs</p></li>
<li><p><strong>max_prop_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Limit to combinations with this proportion in
at least one group.</p></li>
<li><p><strong>min_prop_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – Limit to combinations with this proportion in
in all groups.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Pandas dataframe containing FDR-corrected significance</dt><dd><p>testing results for the distance between paired motifs, between groups</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>res (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.motif_pair_differential_orientation">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">motif_pair_differential_orientation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_pairs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prop_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#motif_pair_differential_orientation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.motif_pair_differential_orientation" title="Link to this definition">¶</a></dt>
<dd><p>Compare the mutual orientation of all motif pairs between groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif_pairs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing the ouptut of
get_motif_pairs.</p></li>
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe containing sequences</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID of group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column in <cite>seqs</cite> containing group IDs</p></li>
<li><p><strong>max_prop_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Limit to combinations with this proportion in
at least one group.</p></li>
<li><p><strong>min_prop_cutoff</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – Limit to combinations with this proportion in
in all groups.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Pandas dataframe containing FDR-corrected significance</dt><dd><p>testing results for the mutual orientation of pairwise combinations
between groups</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>res (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.nmf">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">nmf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#nmf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.nmf" title="Link to this definition">¶</a></dt>
<dd><p>Perform NMF on motif count matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts</strong> (<em>pd.DataFrame</em>) – motif count matrix where rows are sequences and columns
are motifs.</p></li>
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – pandas dataframe containing DNA sequences.</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID for the group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of the column in <cite>seqs</cite> containing group IDs</p></li>
<li><p><strong>n_components</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of components or factors to extract using NMF</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Pandas dataframe of size sequences x factors, containing</dt><dd><p>the contribution of each factor to each sequence.</p>
</dd>
<dt>H (pd.DataFrame): Pandas dataframe of size factors x motifs, containing the</dt><dd><p>contribution of each motif to each factor.</p>
</dd>
<dt>res (pd.DataFrame): Pandas dataframe containing the FDR-corrected significance</dt><dd><p>testing results for factor contribution between groups.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>W (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.scan">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">scan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pthresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#scan"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.scan" title="Link to this definition">¶</a></dt>
<dd><p>Scan a DNA sequence using motifs from a MEME file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Dataframe containing DNA sequences</p></li>
<li><p><strong>meme_file</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to MEME file</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing group IDs</p></li>
<li><p><strong>pthresh</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – p-value cutoff for binding sites</p></li>
<li><p><strong>rc</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Whether to scan the sequence reverse complement as well</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame containing columns ‘MotifID’, ‘SeqID’, ‘start’, ‘end’, ‘strand’.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.motifs.score_sites">
<span class="sig-prename descclassname"><span class="pre">polygraph.motifs.</span></span><span class="sig-name descname"><span class="pre">score_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scores</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/motifs.html#score_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.motifs.score_sites" title="Link to this definition">¶</a></dt>
<dd><p>Calculate the average score of each motif site given base-level importance scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sites</strong> (<em>pd.DataFrame</em>) – Dataframe containing site positions</p></li>
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Dataframe containing sequences</p></li>
<li><p><strong>scores</strong> (<em>np.array</em>) – Numpy array of shape (sequences x length)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Returns</dt><dd><p>sites (pd.DataFrame): ‘sites’ dataframe with an additional columns ‘score’</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.sequence">
<span id="polygraph-sequence-module"></span><h2>polygraph.sequence module<a class="headerlink" href="#module-polygraph.sequence" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.ISM">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">ISM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#ISM"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.ISM" title="Link to this definition">¶</a></dt>
<dd><p>Perform in-silico mutagenesis on given DNA sequence(s)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – A DNA sequence, list of sequences
or dataframe containing sequences in the column “Sequence”.</p></li>
<li><p><strong>drop_ref</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, do not return the original sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A list of all possible single-base mutated sequences</dt><dd><p>derived from the original sequences.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.bleu_similarity">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">bleu_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#bleu_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.bleu_similarity" title="Link to this definition">¶</a></dt>
<dd><p>Calculate the bleu similarity score between two sets of sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of DNA sequences</p></li>
<li><p><strong>reference_seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of DNA sequences</p></li>
<li><p><strong>max_k</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Highest k-mer length for calculation. All k-mers of
length 1 to max_k inclusive will be considered.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.fastsk">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">fastsk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#fastsk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.fastsk" title="Link to this definition">¶</a></dt>
<dd><p>Compute a gapped k-mer kernel matrix for the given sequences using FastSK.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – A DNA sequence, list of sequences
or dataframe containing sequences in the column “Sequence”.</p></li>
<li><p><strong>k</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – k-mer length</p></li>
<li><p><strong>m</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of mismatches allowed</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of shape (n_seqs, n_seqs) containing the gapped k-mer kernel.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.gc">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">gc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.gc" title="Link to this definition">¶</a></dt>
<dd><p>Calculate the GC fraction of a DNA sequence or list of sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – A DNA sequence, list of sequences
or dataframe containing sequences in the column “Sequence”.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The fraction of each sequence comprised of G and C bases.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>, <a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)">float</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.groupwise_mean_edit_dist">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">groupwise_mean_edit_dist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#groupwise_mean_edit_dist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.groupwise_mean_edit_dist" title="Link to this definition">¶</a></dt>
<dd><p>Calculate average edit distances between all groups of sequences</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.kmer_frequencies">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">kmer_frequencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#kmer_frequencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.kmer_frequencies" title="Link to this definition">¶</a></dt>
<dd><p>Get frequencies of all kmers of length k in a sequence or sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – A DNA sequence, list of sequences
or dataframe containing sequences in the column “Sequence”.</p></li>
<li><p><strong>k</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The k-mer length.</p></li>
<li><p><strong>normalize</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to normalize the k-mer counts
by sequence length. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A dataframe of shape (kmers x sequences), containing</dt><dd><p>the frequency of each k-mer in the sequence.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.kmer_positions">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">kmer_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmer</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#kmer_positions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.kmer_positions" title="Link to this definition">¶</a></dt>
<dd><p>Return all the locations of a given k-mer in a DNA sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – the input DNA sequence</p></li>
<li><p><strong>kmer</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – the k-mer for which to search</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a numpy array containing the positions of the kmer</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.min_edit_distance">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">min_edit_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#min_edit_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.min_edit_distance" title="Link to this definition">¶</a></dt>
<dd><p>For each sequence in a list, find the smallest edit distance between that sequence
and a list of reference sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of sequences</p></li>
<li><p><strong>reference_seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of sequences</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>edit distance between each sequence in seqs and its closest reference sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.min_edit_distance_from_reference">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">min_edit_distance_from_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#min_edit_distance_from_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.min_edit_distance_from_reference" title="Link to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>For each sequence in non-reference groups, find the smallest edit distance</dt><dd><p>between that sequence and the sequences in the reference group.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Dataframe containing sequences in column “Sequence”</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID for the group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of the column containing group IDs</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>list of edit distance between each sequence and its closest</dt><dd><p>reference sequence.</p>
</dd>
</dl>
<p>Set to 0 for reference sequences</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>edit (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.sequence.unique_kmers">
<span class="sig-prename descclassname"><span class="pre">polygraph.sequence.</span></span><span class="sig-name descname"><span class="pre">unique_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/sequence.html#unique_kmers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.sequence.unique_kmers" title="Link to this definition">¶</a></dt>
<dd><p>Get all unique kmers of length k that are present in a DNA sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – the input DNA sequence</p></li>
<li><p><strong>k</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – length of k-mers to extract</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>a set containing the unique kmers extracted from</dt><dd><p>the sequence.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#set" title="(in Python v3.10)">set</a>)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.stats">
<span id="polygraph-stats-module"></span><h2>polygraph.stats module<a class="headerlink" href="#module-polygraph.stats" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.stats.groupwise_fishers">
<span class="sig-prename descclassname"><span class="pre">polygraph.stats.</span></span><span class="sig-name descname"><span class="pre">groupwise_fishers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_val</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/stats.html#groupwise_fishers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.stats.groupwise_fishers" title="Link to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Perform Fisher’s exact test for proportions between each non-reference group</dt><dd><p>and the reference group.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em><em>, </em><em>anndata.AnnData</em>) – Pandas dataframe with group IDs
and values to compare, or an AnnData object containing this
dataframe in .obs</p></li>
<li><p><strong>val_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column with values to compare</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID of group to use as reference</p></li>
<li><p><strong>reference_val</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – A specific value whose proportion is to be compared
between groups</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column containing group IDs</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Dataframe containing group proportions and FDR-corrected</dt><dd><p>p-values for each group.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.stats.groupwise_mann_whitney">
<span class="sig-prename descclassname"><span class="pre">polygraph.stats.</span></span><span class="sig-name descname"><span class="pre">groupwise_mann_whitney</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/stats.html#groupwise_mann_whitney"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.stats.groupwise_mann_whitney" title="Link to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Compare the mean values between each non-reference group and the</dt><dd><p>reference group using the Mann-Whitney U test.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em><em>, </em><em>anndata.AnnData</em>) – Pandas dataframe containing group IDs
and values to compare, or an AnnData object containing this dataframe
in .obs</p></li>
<li><p><strong>val_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column with values to compare</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – ID of group to use as reference</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column containing group IDs</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing FDR-corrected p-values for each group.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.stats.kruskal_dunn">
<span class="sig-prename descclassname"><span class="pre">polygraph.stats.</span></span><span class="sig-name descname"><span class="pre">kruskal_dunn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/stats.html#kruskal_dunn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.stats.kruskal_dunn" title="Link to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Compare the mean values between all groups using the Kruskal-Wallis</dt><dd><p>test followed by Dunn’s post-hoc test</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em><em>, </em><em>anndata.AnnData</em>) – Pandas dataframe with group IDs
and values to compare, or an AnnData object containing this dataframe
in .obs</p></li>
<li><p><strong>val_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column with values to compare</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Name of column containing group IDs</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing p-values for both Kruskal-Wallis and Dunn’s test.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#dict" title="(in Python v3.10)">dict</a>)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.utils">
<span id="polygraph-utils-module"></span><h2>polygraph.utils module<a class="headerlink" href="#module-polygraph.utils" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.utils.check_equal_lens">
<span class="sig-prename descclassname"><span class="pre">polygraph.utils.</span></span><span class="sig-name descname"><span class="pre">check_equal_lens</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/utils.html#check_equal_lens"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.utils.check_equal_lens" title="Link to this definition">¶</a></dt>
<dd><p>Given sequences, check whether they are all of equal length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – Either a list of DNA sequences,
or a dataframe containing DNA sequences in the column
“Sequence”.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>whether the sequences are all equal in length.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)">bool</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.utils.get_lens">
<span class="sig-prename descclassname"><span class="pre">polygraph.utils.</span></span><span class="sig-name descname"><span class="pre">get_lens</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/utils.html#get_lens"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.utils.get_lens" title="Link to this definition">¶</a></dt>
<dd><p>Calculate the lengths of given DNA sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – A DNA sequence, list of sequences
or dataframe containing sequences in the column “Sequence”.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>length of each sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)">int</a>, <a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.utils.integer_encode">
<span class="sig-prename descclassname"><span class="pre">polygraph.utils.</span></span><span class="sig-name descname"><span class="pre">integer_encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/utils.html#integer_encode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.utils.integer_encode" title="Link to this definition">¶</a></dt>
<dd><p>Encode DNA sequence(s) as a numpy array of integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – seqs (str, list): A DNA sequence, list of
sequences or dataframe containing sequences in the column “Sequence”.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A 1-D or 2-D array containing the sequences encoded as integers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.utils.make_ids">
<span class="sig-prename descclassname"><span class="pre">polygraph.utils.</span></span><span class="sig-name descname"><span class="pre">make_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/utils.html#make_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.utils.make_ids" title="Link to this definition">¶</a></dt>
<dd><p>Assign a unique index to each row of a dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seqs</strong> (<em>pd.DataFrame</em>) – Pandas dataframe</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Modified database containing unique indices.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>seqs (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.utils.pad_with_Ns">
<span class="sig-prename descclassname"><span class="pre">polygraph.utils.</span></span><span class="sig-name descname"><span class="pre">pad_with_Ns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'both'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/utils.html#pad_with_Ns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.utils.pad_with_Ns" title="Link to this definition">¶</a></dt>
<dd><p>Pads a sequence with Ns at the desired end until it reaches
<cite>seq_len</cite> in length.</p>
<p>If seq_len is not provided, it is set to the length of
the longest sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – DNA sequence, list of sequences
or dataframe containing sequences in the column “Sequence”.</p></li>
<li><p><strong>seq_len</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length upto which to pad each sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Padded sequences of length seq_len</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.utils.reverse_complement">
<span class="sig-prename descclassname"><span class="pre">polygraph.utils.</span></span><span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/utils.html#reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.utils.reverse_complement" title="Link to this definition">¶</a></dt>
<dd><p>Reverse complement DNA sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seqs</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a><em>, </em><em>pd.DataFrame</em>) – seqs (str, list): A DNA sequence, list of
sequences or dataframe containing sequences in the column “Sequence”.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>reverse complemented sequences</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)">list</a>)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph.visualize">
<span id="polygraph-visualize-module"></span><h2>polygraph.visualize module<a class="headerlink" href="#module-polygraph.visualize" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.boxplot">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">boxplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#boxplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.boxplot" title="Link to this definition">¶</a></dt>
<dd><p>Plot boxplot of values in each group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em><em>, </em><em>anndata.AnnData</em>) – Pandas dataframe with group IDs
and values to compare, or an AnnData object containing this
dataframe in .obs</p></li>
<li><p><strong>value_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing values to plot</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing group IDs</p></li>
<li><p><strong>fill_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing additional variable to split each group</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.densityplot">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">densityplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#densityplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.densityplot" title="Link to this definition">¶</a></dt>
<dd><p>Plot density plot of values in each group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em><em>, </em><em>anndata.AnnData</em>) – Pandas dataframe with group IDs
and values to compare, or an AnnData object containing this
dataframe in .obs</p></li>
<li><p><strong>value_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing values to plot</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing group IDs</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.one_nn_frac_plot">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">one_nn_frac_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ad</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#one_nn_frac_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.one_nn_frac_plot" title="Link to this definition">¶</a></dt>
<dd><p>Plot a barplot showing the fraction of points in each group whose
nearest neighbors are reference sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ad</strong> (<em>anndata.AnnData</em>) – AnnData object containing sequence embedding.</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Group to use as reference. This group will
be plotted first.</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column in ad.obs containing group IDs.</p></li>
<li><p><strong>fill_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing additional variable to split each group</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.pca_plot">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">pca_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ad</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0,</span> <span class="pre">1]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_ellipse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#pca_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.pca_plot" title="Link to this definition">¶</a></dt>
<dd><p>Plot PCA embeddings of sequences, colored by group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ad</strong> (<em>anndata.AnnData</em>) – AnnData object containing PCA components.</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing group IDs.</p></li>
<li><p><strong>components</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – PCA components to plot</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – Size of points</p></li>
<li><p><strong>show_ellipse</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Outline each group with an ellipse.</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Group to use as reference. This group will
be plotted first.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.plot_factors_nmf">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_factors_nmf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">H</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#plot_factors_nmf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.plot_factors_nmf" title="Link to this definition">¶</a></dt>
<dd><p>Plot heatmap of contributions of features to NMF factors</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>H</strong> (<em>pd.DataFrame</em>) – Dataframe of shape (factors, features)</p></li>
<li><p><strong>n_features</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of features to cluster</p></li>
<li><p><strong>**kwargs</strong> – Additional arguments to pass to sns.clustermap</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.plot_seqs_nmf">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">plot_seqs_nmf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reorder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#plot_seqs_nmf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.plot_seqs_nmf" title="Link to this definition">¶</a></dt>
<dd><p>Plot stacked barplot of the distribution of NMF factors among sequences,
split by group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<em>pd.DataFrame</em>) – Dataframe of shape n_seqs x (n_factors+1). The last
column should contain group IDs.</p></li>
<li><p><strong>reorder</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.umap_plot">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">umap_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ad</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_ellipse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#umap_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.umap_plot" title="Link to this definition">¶</a></dt>
<dd><p>Plot UMAP embeddings of sequences, colored by group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ad</strong> (<em>anndata.AnnData</em>) – AnnData object containing UMAP embedding.</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column containing group IDs.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – Size of points</p></li>
<li><p><strong>show_ellipse</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – Outline each group with an ellipse.</p></li>
<li><p><strong>reference_group</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Group to use as reference. This group will
be plotted first.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="polygraph.visualize.upset_plot">
<span class="sig-prename descclassname"><span class="pre">polygraph.visualize.</span></span><span class="sig-name descname"><span class="pre">upset_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ad</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/polygraph/visualize.html#upset_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#polygraph.visualize.upset_plot" title="Link to this definition">¶</a></dt>
<dd><p>Plot UpSet plot showing the overlap between features present in
different groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ad</strong> (<em>anndata.AnnData</em>) – AnnData object containing sequence embedding.</p></li>
<li><p><strong>group_col</strong> (<a class="reference external" href="https://docs.python.org/3.10/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Column in ad.obs containing group IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-polygraph">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-polygraph" title="Link to this heading">¶</a></h2>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">polygraph</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Module Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">polygraph package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.classifier">polygraph.classifier module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polygraph-embedding-module">polygraph.embedding module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.evolve">polygraph.evolve module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.input">polygraph.input module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.likelihood">polygraph.likelihood module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.models">polygraph.models module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.motifs">polygraph.motifs module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.sequence">polygraph.sequence module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.stats">polygraph.stats module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.utils">polygraph.utils module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph.visualize">polygraph.visualize module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-polygraph">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="modules.html">polygraph</a><ul>
      <li>Previous: <a href="modules.html" title="previous chapter">polygraph</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2023, Genentech, inc..
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/api/polygraph.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>